clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -1342,10 +1351,17 @@ public void run() {
                         long now = time.milliseconds();
 
                         if (coordinatorUnknown()) {
-                            if (findCoordinatorFuture != null || lookupCoordinator().failed())
+                            if (findCoordinatorFuture != null || lookupCoordinator().failed()) {
```

Comment:
The nested condition is a bit awkward, how about this:

```
// try to find coordinator once if we have not yet done so; otherwise backoff properly
if (findCoordinatorFuture != null) {
    AbstractCoordinator.this.wait(rebalanceConfig.retryBackoffMs);
} else {
    lookupCoordinator();
}

// now findCoordinatorFuture should not be null;
// if it has failed, clear it so that hb thread can try discover again in the next loop in case main thread is busy
if (findCoordinatorFuture.failed()) {
    clearFindCoordinatorFuture();
}
```

Reply:
Sure yeah that's much better